packages/react-i18n/src/babel-plugin.ts

```diff
@@ -0,0 +1,155 @@
+import {TemplateBuilder} from '@babel/template';
+import * as Types from '@babel/types';
+import {NodePath} from '@babel/traverse';
+import stringHash from 'string-hash';
+import path from 'path';
+import fs from 'fs';
+
+interface State {
+  lastImport: NodePath<Types.ImportDeclaration>;
+}
+
+export default function injectWithI18nArguments({
+  types: t,
+  template,
+}: {
+  types: typeof Types;
+  template: TemplateBuilder<Types.ImportDeclaration | Types.ObjectExpression>;
+}) {
+  function fallbackTranslationsImport({id}) {
+    return template(`import ${id} from './translations/en.json';`, {
+      sourceType: 'module',
+    })();
+  }
+
+  function i18nCallExpression({id, fallbackID, decoratorName}) {
+    return template(
+      `${decoratorName}({
+      id: '${id}',
+      fallback: ${fallbackID},
+      async translations(locale) {
+        return await import(/* webpackChunkName: '${id}-i18n' */ \`./translations/$\{locale}.json\`);
+      },
+    })`,
+      {
+        sourceType: 'module',
+        plugins: ['dynamicImport'],
+        preserveComments: true,
+      },
+    )();
+  }
+
+  function addI18nArguments({
+    binding,
+    bindingName,
+    filename,
+    filenameRelative,
+    fallbackID,
+    insertImport,
+  }) {
+    let _hasTranslations;
+
+    for (const refPath of binding.referencePaths) {
+      const callExpression: NodePath = refPath.parentPath;
+
+      if (!callExpression.isCallExpression()) {
+        continue;
+      }
+
+      const args = callExpression.get('arguments');
+      if (args.length !== 0) {
+        continue;
+      }
+
+      if (_hasTranslations == null) {
+        _hasTranslations = hasTranslations(filename);
```

Comment:
Might be worth commenting on why you are pushing this to be later, since it makes the code read a little awkwardly (for good reason, though!)